rfcs/0024-Non-consensus-node-protocol.md

```diff
@@ -0,0 +1,219 @@
+
+A note from the go pubsup library (https://github.com/libp2p/go-libp2p-pubsub) : Subscription is not an instanteneous operation. It may take some time before the subscription is processed by the pubsub main loop and propagated to our peers.
+
+### Staged-ledger-validity
+
+When a node verifies a blockchain snark, it also verifies the ledger state upto a certain sequence of transactions (all the transactions for which there is a ledger proof). The transactions that haven't been completely snarked yet, full nodes apply them to the ledger in staged ledger and verify the staged-ledger-hash that is part of the new block is equal to the one they generated. This however is not possible for non-consensus nodes because they don't have any state apart from the two different block headers. We can make this slightly better for the non-consensus nodes by having the block porducers commit to the transactions being included and having an in-snark check for preventing any possible transaction reversal
+
+This can be done by chaining the ledger hash (of the ledger in the staged ledger) like we currently do for coinbases in the pending coinbase data structure.
+
+The stack operations would now be:
+
+    ``` ocaml
+        module Stack : sig
+            val push : t -> Coinbase.t option -> Frozen_ledger_hash.t -> t
+
+            module Checked : sig
+                val push : t -> Coinbase_data.var option -> Frozen_ledger_hash.var -> (t, 'a) Tick.Checked.t
+            end
+        end
+    ```
+The blockchain snark would then prove that the transactions included in a new ledger proof is infact what is expected by checking that the stack corresponding to the new ledger proof is the same as the one updated in the pending coinbase.
+
+### Shared Libraries
+
+Binary for non-consensus nodes can share the following libraries with the one for full-node:
+
+    1. Gossip net
+    2. Verifier (This needs fixing as it currently loads prover keys as well)
+    3. Snarky libraries (update and reuse snarkette?)
+    4. Best tip prover
+    5. Network pool (for transactions)
+    6. `select` function from the consensus library
+    7. Account, Transaction, user command, blockchain state, consensus state and all its contents 
+
+### Incentivizing-full-nodes
+
+The aim is to run NC nodes natively, on browsers and on mobile platforms using much less resources compared to full nodes. This makes it much more accessible for users to interact with the blockchain. However, NC nodes don't participate in the consensus and full nodes have no direct incentive to gossip block headers or respond to other requests-
+
+    1. Gossip block headers of root and the best tip along with the merkle-list of state hashes. These can be generated without any extra computational cost though
+    2. Respond to merkle-path requests for every new block and new root
+
+So, we need to think about ways to incentivize full nodes. Maybe-
```

Comment:
Don’t think this is too important as we observe similar behavior un-incentivized on bitcoin/ethereum — one reason could be that the full nodes want the network to succeed and so are willing to eat this tiny cost

Reply:
The state data that is currently needed is just the account merkle paths for every new best tip and new `v`th block from the best tip. To some extent we could reduce the number of requests to full nodes by having a mesh of non-consensus nodes each store a set of accounts. The non-consensus nodes would first request from other non-consensus nodes they are connected to before requesting it from a full-node.

Reply:
Another way to enforce this as a public good, if nodes start not doing this - nodes on the transaction gossip net could randomly pretend to be non consensus nodes, and perform requests for merkle paths, and then ban if they don't get a response.

I think for now it would be fine to assume that the cost will be small enough nodes won't put the effort into changing the default behavior, at least initially when the load is low.